File: scripts/adc.c
#include <wiringPi.h>
#include <wiringPiI2C.h>
#include <unistd.h>
#include <stdio.h>
#include <stdlib.h>  // Para a função qsort

// Configurações iniciais
int setup_sensors() {
    int fd;
    
    // Inicializa a biblioteca WiringPi
    if (wiringPiSetup() == -1) {
        printf("Erro ao inicializar o WiringPi\n");
        return -1;
    }

    // Configura o pino GPIO 4 como entrada com pull-up
    pinMode(4, INPUT);
    pullUpDnControl(4, PUD_UP);

    // Inicializa a comunicação I2C com o PCF8591
    fd = wiringPiI2CSetup(0x48);
    if (fd == -1) {
        printf("Erro ao inicializar o PCF8591\n");
        return -1;
    }

    return fd;
}

// Função para comparar dois números (necessária para qsort)
int compare(const void *a, const void *b) {
    return (*(int*)a - *(int*)b);
}

// Leitura analógica (AIN0 e AIN1) com suavização usando média móvel
void read_analog_values(int fd, int *analogValue0, int *analogValue1) {
    int sum0 = 0, sum1 = 0;
    int num_samples = 5;  // Número de amostras para a média móvel (ajuste conforme necessário)
    
    // Realizar múltiplas leituras e somá-las
    for (int i = 0; i < num_samples; i++) {
        sum0 += wiringPiI2CReadReg8(fd, 0x41);  // Leitura de AIN1
        sum1 += wiringPiI2CReadReg8(fd, 0x40);  // Leitura de AIN0
    }

    // Calcular a média das leituras
    *analogValue0 = sum0 / num_samples;
    *analogValue1 = sum1 / num_samples;
}




// Leitura digital (GPIO4)
int read_digital_value() {
    return digitalRead(7); // Leitura digital do GPIO4
}

File: scripts/adc.py
import time
import RPi.GPIO as GPIO
import smbus2 as smbus


# Definir o endereço I2C do PCF8591
PCF8591_ADDR = 0x48  # Endereço I2C do PCF8591 (0x48)
GPIO_PIN = 4         # Pino GPIO 4 para leitura digital

# Inicializar a biblioteca GPIO e configurar os pinos
GPIO.setmode(GPIO.BCM)
GPIO.setup(GPIO_PIN, GPIO.IN, pull_up_down=GPIO.PUD_UP)  # Configura o GPIO4 com pull-up

# Inicializar o barramento I2C
bus = smbus.SMBus(1)

# Função para ler o valor analógico de um canal AIN (0 ou 1)
def read_analog(channel):
    # Enviar o comando para o PCF8591
    bus.write_byte(PCF8591_ADDR, 0x40 | channel)  # 0x40 é o comando de leitura
    time.sleep(0.1)  # Atraso para permitir que o PCF8591 tenha tempo para responder
    value = bus.read_byte(PCF8591_ADDR)  # Lê o valor analógico
    return value

# Loop principal
try:
    while True:
        # Leitura dos canais analógicos AIN0 e AIN1
        analog_value0 = read_analog(0)  # AIN0
        analog_value1 = read_analog(1)  # AIN1

        # Leitura digital do pino GPIO 4
        digital_value = GPIO.input(GPIO_PIN)

        # Exibe os valores lidos
        print(f"AIN0: {analog_value0} | AIN1: {analog_value1} | GPIO 4: {digital_value}")

        # Atraso de 1 segundo entre as leituras
        time.sleep(1)

except KeyboardInterrupt:
    print("\nPrograma interrompido pelo usuário.")
finally:
    GPIO.cleanup()  # Limpar configurações do GPIO ao finalizar o programa
